pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,212 @@
+package dnsnameresolver
+
+import (
+	"context"
+	"strings"
+	"sync"
+	"time"
+
+	networkv1alpha1 "github.com/openshift/api/network/v1alpha1"
+
+	"github.com/sirupsen/logrus"
+
+	discoveryv1 "k8s.io/api/discovery/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/labels"
+	"sigs.k8s.io/controller-runtime/pkg/cache"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+const (
+	ipRemovalGracePeriod = time.Second * 5
+
+	controllerName = "dnsnameresolver_controller"
+)
+
+type Config struct {
+	OperandNamespace         string
+	ServiceName              string
+	DNSNameResolverNamespace string
+}
+
+// reconciler handles the actual DNSNameResolver reconciliation logic in response to events.
+type reconciler struct {
+	config Config
+
+	dnsNameResolverCache cache.Cache
+	client               client.Client
+	res                  *Resolver
+	startResolver        sync.Once
+}
+
+// NewUnmanaged creates and returns a controller that watches DNSNameResolver
+// objects. The controller re-resolves the DNS names which get added to the
+// status of the DNSNameResolver objects. It also removes IP addresses of DNS
+// names, whose TTLs have expired, from the status. This is an unmanaged
+// controller, which means that the manager does not start it.
+func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, []cache.Cache, error) {
+	// Create a new cache for tracking the DNSNameResolver resources in
+	// the DNSNameResolverNamespace.
+	dnsNameResolverCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.DNSNameResolverNamespace: {},
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create a new cache to track the EndpointSlices corresponding to the
+	// CoreDNS pods.
+	corednsEndpointsSliceCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.OperandNamespace: {},
+		},
+		DefaultLabelSelector: labels.SelectorFromSet(labels.Set{
+			discoveryv1.LabelServiceName: config.ServiceName,
+		}),
+	})
```

Comment:
Is there a reason not to have the manager initialize `corednsEndpointsSliceCache` and pass it to `dnsnameresolver.NewUnmanaged` in `dnsnameresolver.Config` and to `dnsnameresolvercrd.New` in `dnsnameresolvercrd.Config`?  It isn't clear to me why one controller should create the cache when both use it, and it seems like it would be simpler if `dnsnameresolver.NewUnmanaged` just did one thing (construct and return a new unmanaged dnsnameresolver controller) rather than two (construct and return a controller *and* a cache).